packages/nx-openapi/src/helpers/generate-openapi-types.ts

```diff
@@ -0,0 +1,119 @@
+ * @returns Generated type contents
+ */
+export async function generateOpenApiTypes(
+    tree: Tree,
+    schemaPath: string,
+    remote = false,
+    configPath?: string
+) {
+    // Parse schema into type definition
+    let contents;
+    if (remote) {
+        console.log('Getting remote schema...');
+        contents = await generateTypesFromRemoteSchema(schemaPath, configPath);
+    } else {
+        console.log('Getting local schema... (use --remote to get remote schema)');
+        contents = await generateTypesFromLocalSchema(tree.root, schemaPath);
+    }
+
+    return contents;
+}
+
+function parseUrlString(url: string): URL {
+    const parsed = new URL(url);
+
+    if (parsed.protocol !== 'https:' && parsed.protocol !== 'http:') {
+        throw new Error(`${parsed.href} is an invalid remote url.`);
+    }
+    return parsed;
+}
+
+/**
+ * Gets the remote schema from an endpoint url. Uses configured authorization or passed in via the user
+ * @param url Remote url to fetch the schema from
+ * @param configPath The path to a local 'redocly' config file. This will be passed into the requests, mainly to specify auth details if required.
+ * @returns a string representation of the remote schema
+ */
+/* v8 ignore start */ // Ignored because these are just wrappers around the ts gen library. No point testing
+export async function generateTypesFromRemoteSchema(url: string, configPath?: string) {
+    const parsedUrl = parseUrlString(url);
+    if (configPath) {
+        const config = await loadConfig({ configPath });
+        console.log('Loaded Config: ', config);
+        console.log('Generating types...');
+        const ast = await openapiTS(parsedUrl, {
+            redocly: config,
+        });
+        return astToString(ast);
+    } else {
+        const ast = await openapiTS(parsedUrl);
+        return astToString(ast);
+    }
+}
+
+/**
+ * Grabs schema data from local directory. The schemaPath is evaluated relative to the root of the template project,
+ * not the root of the generator.
+ * @param rootDir Root directory of the project tree
+ * @param schemaPath Path of the schema relative to the root of the entire project.
+ * @returns a string representation of the schema contents.
+ */
+async function generateTypesFromLocalSchema(rootDir: string, schemaPath: string) {
+    try {
+        console.log('Generating types...');
+        const ast = await openapiTS(`file:///${rootDir}/${schemaPath}`);
+        return astToString(ast);
+    } catch (e) {
+        throw new Error(
+            `Failed to generate local file at path ${rootDir}/${schemaPath} (did you mean to pass --remote?)` +
+                e
+        );
+    }
+}
+/* v8 ignore end */
+
+/**
+ * Copies the original schema from the source to newly generated client
+ */
+export async function copySchema(tree: Tree, name: string, schemaPath: string, remote?: boolean) {
+    let schemaBuffer;
+    if (remote) {
+        const response = await fetch(schemaPath);
```

Comment:
Will this work if the schema is a private one that require authentication?

Reply:
The authentication that you mentioned about is for `generateTypesFromRemoteSchema()` which utilises redocly to download the schema. Here, in this `copySchema()`, we use `fetch` to download the schema but I don't see any configuration for a private endpoint.

Reply:
Added a bug for this in our board: https://aligent.atlassian.net/browse/MI-203